openedx_learning/apps/authoring/collections/api.py

```diff
@@ -70,6 +89,95 @@ def update_collection(
+    created_by: int | None = None,
+) -> int:
+    """
+    Adds a QuerySet of PublishableEntities to a QuerySet of Collections.
+
+    Records are created in bulk, and so integrity errors are deliberately ignored: they indicate that the entity(s)
+    have already been added to the collection(s).
+
+    Returns the number of entities added (including any that already exist).
+    """
+    collection_entities = []
+    entity_ids = entities_qset.values_list("pk", flat=True)
+    collection_ids = collections_qset.values_list("pk", flat=True)
+
+    for collection_id in collection_ids:
+        for entity_id in entity_ids:
+            collection_entities.append(
+                CollectionPublishableEntity(
+                    collection_id=collection_id,
+                    entity_id=entity_id,
+                    created_by_id=created_by,
+                )
+            )
+
+    created = []
+    if collection_entities:
+        created = CollectionPublishableEntity.objects.bulk_create(
+            collection_entities,
+            ignore_conflicts=True,
+        )
+
+        # Update the modified date for all the provided Collections.
+        # (Ignoring conflicts means we don't know which ones were modified.)
+        collections_qset.update(modified=timezone.now())
+
+    return len(created)
+
+
+def remove_from_collections(
+    collections_qset: QuerySet[Collection],
+    entities_qset: QuerySet[PublishableEntity],
+) -> int:
+    """
+    Removes a QuerySet of PublishableEntities from a QuerySet of Collections.
+
+    PublishableEntities are deleted from each Collection, in bulk.
+
+    Collections which had entities removed are marked with modified=now().
+
+    Returns the total number of entities deleted.
+    """
+    total_deleted = 0
+    entity_ids = entities_qset.values_list("pk", flat=True)
+    collection_ids = collections_qset.values_list("pk", flat=True)
+    modified_collection_ids = []
+
+    for collection_id in collection_ids:
+        num_deleted, _ = CollectionPublishableEntity.objects.filter(
+            collection_id=collection_id,
+            entity__in=entity_ids,
+        ).delete()
+
+        if num_deleted:
+            modified_collection_ids.append(collection_id)
+
+        total_deleted += num_deleted
+
+    # Update the modified date for the affected Collections
+    Collection.objects.filter(id__in=modified_collection_ids).update(modified=timezone.now())
+
+    return total_deleted
+
+
+def get_entity_collections(learning_package_id: int, entity_key: str) -> QuerySet[Collection]:
+    """
+    Get all collections in the given learning package which contain this entity.
+
+    Only enabled collections are returned.
+    """
+    entity = publishing_api.get_publishable_entity_by_key(
+        learning_package_id,
+        key=entity_key,
+    )
+    return entity.collections.filter(enabled=True).order_by("pk")
+
+
 def get_learning_package_collections(learning_package_id: int) -> QuerySet[Collection]:
```

Comment:
TODO: rename `get_learning_package_collections` to `get_collections`, cf https://github.com/openedx/edx-platform/pull/35321#discussion_r1734711274